File: backend/igc_parser.py
"""Parser for IGC flight recorder files.

The IGC format is defined by the FAI (Fédération Aéronautique Internationale).
We only need the records most relevant for analysis: H (header), B (GPS fix),
and a couple of optional ones (HFDTE date, HFPLT pilot, HFGTY glider type).
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date, datetime, time, timezone
from pathlib import Path
from typing import Iterable


@dataclass
class Fix:
    timestamp: datetime
    latitude: float
    longitude: float
    pressure_altitude: int
    gps_altitude: int
    valid: bool


@dataclass
class IGCFile:
    flight_date: date | None = None
    pilot: str | None = None
    aircraft_type: str | None = None
    aircraft_id: str | None = None
    competition_id: str | None = None
    fixes: list[Fix] = field(default_factory=list)


def _parse_lat(raw: str) -> float:
    """Convert IGC latitude (DDMMmmmN/S) to signed decimal degrees."""
    deg = int(raw[0:2])
    minutes = int(raw[2:7]) / 1000.0
    value = deg + minutes / 60.0
    return -value if raw[7] in ("S", "s") else value


def _parse_lon(raw: str) -> float:
    """Convert IGC longitude (DDDMMmmmE/W) to signed decimal degrees."""
    deg = int(raw[0:3])
    minutes = int(raw[3:8]) / 1000.0
    value = deg + minutes / 60.0
    return -value if raw[8] in ("W", "w") else value


def _parse_header_value(line: str) -> str:
    """Return the value after the first colon, or after the 5-char prefix."""
    if ":" in line:
        return line.split(":", 1)[1].strip()
    return line[5:].strip()


def parse_igc(lines: Iterable[str]) -> IGCFile:
    igc = IGCFile()
    flight_date = None
    last_seconds = -1
    day_offset = 0  # seconds, to handle UTC midnight rollover

    for raw_line in lines:
        line = raw_line.rstrip("\r\n")
        if not line:
            continue

        rec_type = line[0]

        if rec_type == "H":
            # Header records
            tlc = line[2:5].upper() if len(line) >= 5 else ""
            if tlc == "DTE":
                # HFDTEDDMMYY or HFDTEDATE:DDMMYY,NN
                digits = "".join(c for c in line if c.isdigit())
                if len(digits) >= 6:
                    dd, mm, yy = int(digits[0:2]), int(digits[2:4]), int(digits[4:6])
                    year = 2000 + yy if yy < 80 else 1900 + yy
                    try:
                        flight_date = date(year, mm, dd)
                        igc.flight_date = flight_date
                    except ValueError:
                        pass
            elif tlc == "PLT":
                igc.pilot = _parse_header_value(line)
            elif tlc == "GTY":
                igc.aircraft_type = _parse_header_value(line)
            elif tlc == "GID":
                igc.aircraft_id = _parse_header_value(line)
            elif tlc == "CID":
                igc.competition_id = _parse_header_value(line)

        elif rec_type == "B" and len(line) >= 35:
            # B HHMMSS DDMMmmm[N/S] DDDMMmmm[E/W] A PPPPP GGGGG
            try:
                hh = int(line[1:3])
                mm = int(line[3:5])
                ss = int(line[5:7])
                lat = _parse_lat(line[7:15])
                lon = _parse_lon(line[15:24])
                fix_valid = line[24] == "A"
                pressure_alt = int(line[25:30])
                gps_alt = int(line[30:35])
            except ValueError:
                continue

            seconds = hh * 3600 + mm * 60 + ss
            if seconds < last_seconds - 60:
                # UTC day rollover during the flight.
                day_offset += 86400
            last_seconds = seconds
            base_date = flight_date or date(1970, 1, 1)
            base_dt = datetime.combine(base_date, time(0, 0), tzinfo=timezone.utc)
            timestamp = base_dt.fromtimestamp(
                base_dt.timestamp() + seconds + day_offset, tz=timezone.utc
            )

            igc.fixes.append(
                Fix(
                    timestamp=timestamp,
                    latitude=lat,
                    longitude=lon,
                    pressure_altitude=pressure_alt,
                    gps_altitude=gps_alt,
                    valid=fix_valid,
                )
            )

    return igc


def parse_igc_file(path: str | Path) -> IGCFile:
    with open(path, "r", encoding="latin-1") as f:
        return parse_igc(f)


def parse_igc_bytes(data: bytes) -> IGCFile:
    text = data.decode("latin-1")
    return parse_igc(text.splitlines())

File: frontend/src/utils/igcParser.ts
export interface IgcFix {
  timestampMs: number;       // unix epoch ms (UTC)
  timestampIso: string;      // ISO 8601 with Z
  latitude: number;
  longitude: number;
  pressureAltitude: number;
  gpsAltitude: number;
  valid: boolean;
}

export interface IgcFile {
  flightDate: string | null; // YYYY-MM-DD
  pilot: string | null;
  aircraftType: string | null;
  aircraftId: string | null;
  competitionId: string | null;
  fixes: IgcFix[];
}

function parseLat(raw: string): number {
  const deg = parseInt(raw.slice(0, 2), 10);
  const minutes = parseInt(raw.slice(2, 7), 10) / 1000.0;
  const value = deg + minutes / 60.0;
  const hemi = raw[7];
  return hemi === "S" || hemi === "s" ? -value : value;
}

function parseLon(raw: string): number {
  const deg = parseInt(raw.slice(0, 3), 10);
  const minutes = parseInt(raw.slice(3, 8), 10) / 1000.0;
  const value = deg + minutes / 60.0;
  const hemi = raw[8];
  return hemi === "W" || hemi === "w" ? -value : value;
}

function parseHeaderValue(line: string): string {
  const i = line.indexOf(":");
  if (i >= 0) return line.slice(i + 1).trim();
  return line.slice(5).trim();
}

/** Format Date as ISO 8601 with seconds precision and `+00:00` suffix, matching Python's datetime.isoformat(). */
function toIsoUtcSeconds(ms: number): string {
  const d = new Date(ms);
  const pad = (n: number, w = 2) => String(n).padStart(w, "0");
  return (
    `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}` +
    `T${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}+00:00`
  );
}

export function parseIgcText(text: string): IgcFile {
  const igc: IgcFile = {
    flightDate: null,
    pilot: null,
    aircraftType: null,
    aircraftId: null,
    competitionId: null,
    fixes: [],
  };

  let flightDateMs: number | null = null; // UTC midnight of flight date
  let lastSeconds = -1;
  let dayOffsetS = 0;

  const lines = text.split(/\r?\n/);
  for (const raw of lines) {
    const line = raw.replace(/[\r\n]+$/, "");
    if (!line) continue;

    const rec = line[0];

    if (rec === "H") {
      const tlc = line.length >= 5 ? line.slice(2, 5).toUpperCase() : "";
      if (tlc === "DTE") {
        const digits = line.replace(/\D/g, "");
        if (digits.length >= 6) {
          const dd = parseInt(digits.slice(0, 2), 10);
          const mm = parseInt(digits.slice(2, 4), 10);
          const yy = parseInt(digits.slice(4, 6), 10);
          const year = yy < 80 ? 2000 + yy : 1900 + yy;
          // Validate
          const candidate = Date.UTC(year, mm - 1, dd);
          const back = new Date(candidate);
          if (
            back.getUTCFullYear() === year &&
            back.getUTCMonth() === mm - 1 &&
            back.getUTCDate() === dd
          ) {
            flightDateMs = candidate;
            igc.flightDate = `${year}-${String(mm).padStart(2, "0")}-${String(dd).padStart(2, "0")}`;
          }
        }
      } else if (tlc === "PLT") {
        igc.pilot = parseHeaderValue(line);
      } else if (tlc === "GTY") {
        igc.aircraftType = parseHeaderValue(line);
      } else if (tlc === "GID") {
        igc.aircraftId = parseHeaderValue(line);
      } else if (tlc === "CID") {
        igc.competitionId = parseHeaderValue(line);
      }
    } else if (rec === "B" && line.length >= 35) {
      const hh = parseInt(line.slice(1, 3), 10);
      const mm = parseInt(line.slice(3, 5), 10);
      const ss = parseInt(line.slice(5, 7), 10);
      if (!Number.isFinite(hh) || !Number.isFinite(mm) || !Number.isFinite(ss)) continue;

      const lat = parseLat(line.slice(7, 15));
      const lon = parseLon(line.slice(15, 24));
      if (!Number.isFinite(lat) || !Number.isFinite(lon)) continue;

      const fixValid = line[24] === "A";
      const pressureAlt = parseInt(line.slice(25, 30), 10);
      const gpsAlt = parseInt(line.slice(30, 35), 10);
      if (!Number.isFinite(pressureAlt) || !Number.isFinite(gpsAlt)) continue;

      const seconds = hh * 3600 + mm * 60 + ss;
      if (seconds < lastSeconds - 60) {
        dayOffsetS += 86400;
      }
      lastSeconds = seconds;

      const baseMs = flightDateMs ?? Date.UTC(1970, 0, 1);
      const ms = baseMs + (seconds + dayOffsetS) * 1000;

      igc.fixes.push({
        timestampMs: ms,
        timestampIso: toIsoUtcSeconds(ms),
        latitude: lat,
        longitude: lon,
        pressureAltitude: pressureAlt,
        gpsAltitude: gpsAlt,
        valid: fixValid,
      });
    }
  }

  return igc;
}

/** Read a File as latin-1 text and parse it. */
export async function parseIgcFile(file: File | Blob): Promise<IgcFile> {
  // Latin-1 keeps every byte 1:1, matching the Python parser's `latin-1` decode.
  const buf = await file.arrayBuffer();
  const bytes = new Uint8Array(buf);
  // Build a string by char codes — equivalent to latin-1 decode without a TextDecoder dependency.
  let text = "";
  // Chunk to avoid call-stack issues on very large files.
  const CHUNK = 0x8000;
  for (let i = 0; i < bytes.length; i += CHUNK) {
    text += String.fromCharCode.apply(null, Array.from(bytes.subarray(i, i + CHUNK)));
  }
  return parseIgcText(text);
}
